File: combination_sum_ii.cpp
/*
Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

Each number in C may only be used once in the combination.

Note:
All numbers (including target) will be positive integers.
Elements in a combination (a1, a2, … , ak) must be in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
The solution set must not contain duplicate combinations.
For example, given candidate set 10,1,2,7,6,1,5 and target 8, 
A solution set is: 
[1, 7] 
[1, 2, 5] 
[2, 6] 
[1, 1, 6] 
*/

#include "header.h"

void search(vector<int>& candidates,int curr_idx, int target,int curr_sum, vector<int>& temp, vector<vector<int> >& res){
    int sz = candidates.size();
    if(curr_sum == target){
        res.push_back(temp);
        return;
    }
    if(curr_idx>=sz || curr_sum>target) return;
    vector<int> orig = temp;
    int x = curr_idx+1;
    int orig_sum = curr_sum;
    while(x<sz && candidates[x] == candidates[x-1]) x++;
    for(int i=curr_idx; i<x; i++){
        temp.push_back(candidates[curr_idx]);
        curr_sum += candidates[curr_idx];
        search(candidates,x,target,curr_sum,temp,res);
    }
    temp = orig;
    search(candidates,x,target,orig_sum,temp,res);
}

vector<vector<int> > combinationSum2(vector<int>& candidates, int target){
    int sz = candidates.size();
    vector<vector<int> > res;
    if(sz == 0) return res;
    sort(candidates.begin(), candidates.end());
    vector<int> temp;
    int curr_sum = 0;
    search(candidates,0,target,curr_sum,temp,res);
    return res;
}

int main(){
    //int arr[] = {10,1,2,7,6,1,5};
    //int arr[] = {2,1,3,1,4};
    int arr[] = {1,1,2,5,6,7,10};
    //int arr[] = {1,1};
    vector<int> candidates(arr,arr+sizeof(arr)/sizeof(arr[0]));
    vector<vector<int> > res = combinationSum2(candidates,10);
//    vector<vector<int> > res = combinationSum2(candidates,8);
    for(int i=0; i<res.size(); i++){
        for(int j=0; j<res[i].size(); j++)
            printf("%d ", res[i][j]);
        printf("\n");
    }
    return 0;
}

File: combination_sum_ii.py
import pdb

class Solution:
    def comb_helper(self,candidates,target,res,s,idx,curr_sum):
        if curr_sum == target:
            res.append(s[:])
            return
        if idx>=len(candidates):
            return
        if candidates[idx]+curr_sum > target:
            return
        temp = idx
        while idx+1<len(candidates) and candidates[idx+1] == candidates[idx]:
            idx+=1
        self.comb_helper(candidates,target,res,s,idx+1,curr_sum)
        s.append(candidates[temp])
        self.comb_helper(candidates,target,res,s,temp+1,curr_sum+candidates[temp])
        s.pop()

    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum2(self, candidates, target):
        if candidates == []:
            return []
        res = []
        s = []
        candidates = sorted(candidates)
        self.comb_helper(candidates,target,res,s,0,0)
        return res

if __name__ == "__main__":
    #print Solution().combinationSum2([10, 1, 2, 7, 6, 1, 5], 8)
    print Solution().combinationSum2([1], 1)
